### PR TITLE
Add ShoppingList Reference Resolver helpers/utilities.

### PR DESCRIPTION
**Pull request by leungkinghin**
#### Relevant Issues
https://github.com/commercetools/commercetools-sync-java/issues/594

#### Hints for Review
- LineItemDraft variant selection (https://docs.commercetools.com/http-api-projects-shoppingLists#lineitemdraft-product-variant-selection) will be provided as SKU.
- customer field will be supplied as ResourceIdentifier with key
- custom field will be supplied as ResourceIdentifier with key
- TextLineItemDraft -> custom field will be supplied as ResourceIdentifier with key
- LineItemDraft -> custom field will be supplied as ResourceIdentifier with key

### CODE REVIEW

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/LineItemReferenceResolverTest.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    @BeforeEach
+    void setup() {
+        typeService = getMockTypeService();
+
+        final ShoppingListSyncOptions syncOptions = ShoppingListSyncOptionsBuilder.of(mock(SphereClient.class)).build();
+        referenceResolver = new LineItemReferenceResolver(syncOptions, typeService);
+    }
+
+    @Test
+    void resolveReferences_WithNonNullIdOnCustomTypeResId_ShouldResolveCustomTypeReference() {
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/shoppinglists/helpers/LineItemReferenceResolverTest.java:48]: Does it resolve anything in this test? I see that `lineItemDraft` is the same as `resolvedDraft`.

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/LineItemReferenceResolverTest.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        // assertion
+        assertThat(referenceResolver.resolveReferences(lineItemDraft))
+            .hasFailedWithThrowableThat()
+            .isExactlyInstanceOf(SphereException.class)
+            .hasMessageContaining("CTP error on fetch");
+    }
+
+    @Test
+    void resolveReferences_WithNonExistentCustomType_ShouldCompleteExceptionally() {
+        final String customTypeKey = "customTypeKey";
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/shoppinglists/helpers/LineItemReferenceResolverTest.java:116]: // preparation

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/LineItemReferenceResolverTest.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        ;
+        assertThat(referenceResolver.resolveReferences(lineItemDraft))
+            .hasFailedWithThrowableThat()
+            .isExactlyInstanceOf(ReferenceResolutionException.class)
+            .hasMessage(expectedMessageWithCause);
+    }
+
+    @Test
+    void resolveReferences_WithEmptyKeyOnCustomTypeResId_ShouldCompleteExceptionally() {
+        final CustomFieldsDraft customFieldsDraft = CustomFieldsDraft
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/shoppinglists/helpers/LineItemReferenceResolverTest.java:142]: // preparation

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+        typeService = getMockTypeService();
+        customerService = getMockCustomerService();
+
+        final ShoppingListSyncOptions syncOptions = ShoppingListSyncOptionsBuilder.of(mock(SphereClient.class)).build();
+        referenceResolver = new ShoppingListReferenceResolver(syncOptions, customerService, typeService);
+    }
+
+    @Test
+    void resolveCustomTypeReference_WithNonNullIdOnCustomTypeResId_ShouldResolveCustomTypeReference() {
+        final CustomFieldsDraft customFieldsDraft = CustomFieldsDraft
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java:69]: // preparation

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+
+        // assertion
+        assertThat(resolvedDraft.getCustom()).isNotNull();
+        assertThat(resolvedDraft.getCustom().getType().getId()).isEqualTo("typeId");
+
+    }
+
+    @Test
+    void resolveCustomTypeReference_WithNonNullKeyOnCustomTypeResId_ShouldResolveCustomTypeReference() {
+        final String customTypeId = "customTypeId";
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java:91]: // preparation

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+            .toCompletableFuture().join();
+
+        // assertion
+        assertThat(resolvedDraft.getCustom()).isNotNull();
+        assertThat(resolvedDraft.getCustom().getType().getId()).isEqualTo("typeId");
+
+    }
+
+    @Test
+    void resolveCustomTypeReference_WithNonNullKeyOnCustomTypeResId_ShouldResolveCustomTypeReference() {
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java:90]: Should it be WithNonNullId for this test?

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+        ;
+        assertThat(resolvedDraftCompletionStage)
+            .hasFailedWithThrowableThat()
+            .isExactlyInstanceOf(ReferenceResolutionException.class)
+            .hasMessage(expectedMessageWithCause);
+    }
+
+    @Test
+    void resolveCustomTypeReference_WithEmptyKeyOnCustomTypeResId_ShouldCompleteExceptionally() {
+        final CustomFieldsDraft customFieldsDraft = CustomFieldsDraft
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java:172]: // preparation

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+        final ShoppingListDraftBuilder referencesResolvedDraft =
+            referenceResolver.resolveCustomerReference(draftBuilder).toCompletableFuture().join();
+
+        // assertion
+        assertThat(referencesResolvedDraft.getCustom()).isNull();
+    }
+
+    @Test
+    void resolveCustomerReference_WithExceptionOnCustomerGroupFetch_ShouldNotResolveReference() {
+        final ShoppingListDraftBuilder draftBuilder =
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java:232]: // preparation

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+            ShoppingListDraftBuilder
+                .of(LocalizedString.of(Locale.ENGLISH, "NAME"))
+                .customer(ResourceIdentifier.ofKey("anyKey"))
+                .description(LocalizedString.of(Locale.ENGLISH, "DESCRIPTION"));
+
+        final CompletableFuture<Optional<String>> futureThrowingSphereException = new CompletableFuture<>();
+        futureThrowingSphereException.completeExceptionally(new SphereException("CTP error on fetch"));
+        when(customerService.fetchCachedCustomerId(anyString())).thenReturn(futureThrowingSphereException);
+
+        assertThat(referenceResolver.resolveCustomerReference(draftBuilder).toCompletableFuture())
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java:242]: // test and assertion

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+        // test and assertion
+        assertThat(referenceResolver.resolveCustomerReference(draftBuilder).toCompletableFuture())
+            .hasNotFailed()
+            .isCompletedWithValueMatching(resolvedDraft ->
+                Objects.equals(resolvedDraft.getCustomer(), draftBuilder.getCustomer()));
+    }
+
+    @Test
+    void resolveReferences_WithoutReferences_ShouldNotResolveReferences() {
+        final ShoppingListDraft shoppingListDraft = ShoppingListDraftBuilder
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java:321]: // preparation, test and assertions

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+            .resolveReferences(shoppingListDraft)
+            .toCompletableFuture()
+            .join();
+
+        assertThat(resolvedDraft).isEqualTo(shoppingListDraft);
+    }
+
+    @Test
+    void resolveReferences_WithAllValidFieldsAndReferences_ShouldResolveReferences() {
+        final CustomFieldsDraft customFieldsDraft = CustomFieldsDraft
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java:336]: // preparation, test and assertions

Comment (lojzatran) [reply src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java:336]: Can you check in general all the test files and add those sections where missing?

File: src/test/java/com/commercetools/sync/shoppinglists/utils/ShoppingListsReferenceResolutionUtilsTest.java [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+            assertThat(draft.getTextLineItems()).containsExactly(
+                TextLineItemDraftBuilder
+                    .of(LocalizedString.ofEnglish("textLineItemName"), 0L)
+                    .custom(CustomFieldsDraft.ofTypeKeyAndJson("customTypeKey", emptyMap()))
+                    .build());
+        });
+    }
+
+    @Test
+    void mapToShoppingListDrafts_WithNonExpandedReferences_ShouldReturnResourceIdentifiersWithKeys() {
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/shoppinglists/utils/ShoppingListsReferenceResolutionUtilsTest.java:99]: If you don't expand, how can it know the keys?

**Review by lojzatran — CHANGES_REQUESTED** [review]

File: src/test/java/com/commercetools/sync/shoppinglists/utils/ShoppingListsReferenceResolutionUtilsTest.java (hunk 12 above)
Comment (ahmetoz) [reply src/test/java/com/commercetools/sync/shoppinglists/utils/ShoppingListsReferenceResolutionUtilsTest.java]: You are right, renamed the test name: https://github.com/commercetools/commercetools-sync-java/pull/602/commits/2fdb1c3a8a1d24a53fa441040e53fd85c841c0d6

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java (hunk 11 above)
Comment (ahmetoz) [reply src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java:312]: you are right, I am not a fan of those comments, all of them removed:  https://github.com/commercetools/commercetools-sync-java/pull/602/commits/992d11e4290aaad05188839eff64777e4de88d0a

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java (hunk 6 above)
Comment (ahmetoz) [reply src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java]: check this name, please: https://github.com/commercetools/commercetools-sync-java/pull/602/commits/e895c7a9806d27c3c07c712b42e3fbd894ff318c

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/LineItemReferenceResolverTest.java (hunk 1 above)
Comment (ahmetoz) [reply src/test/java/com/commercetools/sync/shoppinglists/helpers/LineItemReferenceResolverTest.java]: you are right, check with this name, please: https://github.com/commercetools/commercetools-sync-java/pull/602/commits/197871b54f23763c3218e14ea19e59aba4a88d1c

File: src/test/java/com/commercetools/sync/products/ProductSyncMockUtils.java [hunk 13]
```diff
@@ -505,4 +506,18 @@ public static CustomObjectService getMockCustomObjectService(@Nonnull final Stri
...
     }
+
+    /**
+     * Creates a mock {@link Customer} with the supplied {@code id} and {@code key}.
+     *
+     * @param id  the id of the created mock {@link Customer}.
+     * @param key the key of the created mock {@link CustomerGroup}.
+     * @return a mock customerGroup with the supplied id and key.
+     */
+    public static Customer getMockCustomer(final String id, final String key) {
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/test/java/com/commercetools/sync/products/ProductSyncMockUtils.java]: moved to a better place: https://github.com/commercetools/commercetools-sync-java/pull/602/commits/a0a09e9664b08116450602f6af669e39dd166a62

File: src/main/java/com/commercetools/sync/shoppinglists/utils/ShoppingListReferenceResolutionUtils.java [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+import static com.commercetools.sync.commons.utils.CustomTypeReferenceResolutionUtils.mapToCustomFieldsDraft;
+import static com.commercetools.sync.commons.utils.SyncUtils.getResourceIdentifierWithKey;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Util class which provides utilities that can be used when syncing resources from a source commercetools project
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/shoppinglists/utils/ShoppingListReferenceResolutionUtils.java]: https://github.com/commercetools/commercetools-sync-java/pull/602/commits/447d01600d7d078ba3514b5f07dd64455d0fa063

File: src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java (hunk 11 above)
Comment (ahmetoz) [reply src/test/java/com/commercetools/sync/shoppinglists/helpers/ShoppingListReferenceResolverTest.java:312]: note: to help you, I resolved similar comments. Let's discuss it here.

**Review by ahmetoz — APPROVED** [review]

**Review by lojzatran — APPROVED** [review]

File: src/test/java/com/commercetools/sync/shoppinglists/utils/ShoppingListsReferenceResolutionUtilsTest.java [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+
+    @Test
+    void mapToShoppingListDrafts_WithExpandedReferences_ShouldReturnResourceIdentifiersWithKeys() {
+        final Type mockCustomType = getTypeMock(UUID.randomUUID().toString(), "customTypeKey");
+        final Customer mockCustomer = getMockCustomer(UUID.randomUUID().toString(), "customerKey");
+        final ProductVariant mockProductVariant = mock(ProductVariant.class);
+        when(mockProductVariant.getSku()).thenReturn("variant-sku");
+
+        final List<ShoppingList> mockShoppingLists = new ArrayList<>();
+        mockShoppingLists.add(null);
```
Comment (leungkinghin) [inline src/test/java/com/commercetools/sync/shoppinglists/utils/ShoppingListsReferenceResolutionUtilsTest.java:45]: what's the purpose to add a null object into the list?

Comment (ahmetoz) [reply src/test/java/com/commercetools/sync/shoppinglists/utils/ShoppingListsReferenceResolutionUtilsTest.java:45]: To ensure that null drafts filtered out.

**Pull request merged**